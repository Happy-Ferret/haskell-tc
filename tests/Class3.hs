{-# LANGUAGE ExplicitForAll #-}
module Class3 where

-- Errors:
--   Could not deduce (Show b) from context.
--     Kept contains a skolem variable
--     Example:
--       fn :: Class a => a -> String
--       err :: a -> String
--       err a = fn a
--   No instance for (Show meta_var_10)
--     Kept contains a meta variable
--     Example:
--       fn :: Class a => a -> String
--       err = fn undefined
--       err x = show (read x) -- with no defaulting
--   context too weak
--     Context is not null.

data String

-- class Show a where
--
-- show :: Show a => a -> String
-- show x = show x

class Super a
class Super a => Sub a
-- class Sub a
instance (Super a, Super b) => Super (a,b)

super :: Super a => a -> String
super x = super x

-- sub :: Sub a => a -> String
-- sub x = super x

-- byInst :: (Super a, Super b) => a -> b -> String
byInst a b = super (a,b)

-- show' :: (Show a, Show b)  => a -> b -> String
-- show' a b = show' a b

-- Could not deduce (Show b) from context.
-- show2 :: Show a => a -> b -> String
-- show2 a b = show2 b a

-- undefined :: a -> b
-- undefined x = undefined x

-- No instance for (Show meta_var_10)
-- err3 :: a -> String
-- err3 x = show' x (undefined x)

-- context too weak
-- bad :: a -> String
-- bad x = show x

-- s2s :: String -> String
-- s2s x = s2s x

-- Signature too general
-- unify 'String' with skolem_a
-- s2any :: String -> a
-- s2any x = s2s x

-- show' x = show x

{-
add assumption: show :: forall a. Show a => a -> String
show' :: _A_
x     :: _B_
_A_ = _B_ -> _C_
instantiate (forall a. Show a => a -> String) => _D_ -> String
add predicate: Show _D_
_B_ = _D_
_C_ = String
_A_ = _D_ -> String
generalize: _A_ => _D_ -> String => forall a. a -> String
Add predicates: forall a. Show a => a -> String
-}
