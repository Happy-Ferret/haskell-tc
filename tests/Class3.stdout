Bindings:
:: âˆ€ sk_21_a. .Super sk_21_a â‡’ sk_21_a â†’  String
  [4msuper x = super x[0m
:: sk_21_a
  super [4mx[0m = super x
:: âˆ€ v5 v6. (.Super v6, .Super v5) â‡’ v5 â†’  v6 â†’  String
  [4mbyInst a b = super (a,b)[0m
:: v5
  byInst [4ma[0m b = super (a,b)
:: v6
  byInst a [4mb[0m = super (a,b)

Proofs:
coercion: (âˆ€ a. .Super a â‡’ a â†’  String) @ sk_21_a
  super x = [4msuper[0m x
coercion: (âˆ€ a. .Super a â‡’ a â†’  String) @ (v5
                                          ,v6)
  byInst a b = [4msuper[0m (a,b)
